tests/test_rag_adapter/test_rag_end_to_end.py

```diff
@@ -0,0 +1,320 @@
+
+    # Create a simple RAG adapter that uses our mocked LLM
+    class TestRAGAdapter(GenericRAGAdapter):
+        def __init__(self, vector_store, rag_config):
+            self.vector_store = vector_store
+            self.config = rag_config
+            from gepa.adapters.generic_rag_adapter.evaluation_metrics import RAGEvaluationMetrics
+            from gepa.adapters.generic_rag_adapter.rag_pipeline import RAGPipeline
+
+            self.rag_pipeline = RAGPipeline(vector_store, simple_rag_lm, rag_config)
+            self.evaluator = RAGEvaluationMetrics()
+
+    # Create RAG configuration
+    rag_config = {"retrieval_strategy": "similarity", "top_k": 3, "retrieval_weight": 0.4, "generation_weight": 0.6}
+
+    print("Initializing GenericRAGAdapter with AI/ML knowledge base...")
+
+    # Create the RAG adapter with our mocked LLM
+    adapter = TestRAGAdapter(vector_store=mock_chromadb_store, rag_config=rag_config)
+
+    # Use subset for faster testing
+    trainset = sample_ai_ml_dataset[:2]  # First 2 examples for training
+    valset = sample_ai_ml_dataset[2:3]  # Third example for validation
+
+    # Initial seed candidate with basic RAG prompts
+    seed_candidate = {"answer_generation": "Answer the question '{query}' using the provided context: {context}"}
+
+    # 2. Execution: Run the core RAG optimization logic
+    print("Running GEPA optimization process on RAG system...")
+    gepa_result = gepa.optimize(
+        seed_candidate=seed_candidate,
+        trainset=trainset,
+        valset=valset,
+        adapter=adapter,
+        max_metric_calls=5,  # Small number for fast testing
+        reflection_lm=simple_reflection_lm,
+        display_progress_bar=True,
+    )
+
+    # 3. Assertions: Verify the optimization completed successfully
+    assert gepa_result is not None
+    assert hasattr(gepa_result, "best_candidate")
+    assert hasattr(gepa_result, "val_aggregate_scores")
+
+    best_config = gepa_result.best_candidate
+    best_score = gepa_result.val_aggregate_scores[0]  # First (best) score
+
+    # Basic validation of results
+    assert isinstance(best_config, dict)
+    assert len(best_config) > 0
+    assert "answer_generation" in best_config
+    assert isinstance(best_score, int | float)
+    assert best_score >= 0
+
+    # Verify the prompt was actually optimized (should be different from seed)
+    optimized_prompt = best_config["answer_generation"]
+    seed_prompt = seed_candidate["answer_generation"]
+
+    # The optimized prompt should either be the same (if no improvement found)
+    # or different (if GEPA found a better version)
+    assert isinstance(optimized_prompt, str)
+    assert len(optimized_prompt) > 0
+
+    # Verify GEPA completed metric calls and evaluations
+    assert gepa_result.total_metric_calls > 0
+    assert gepa_result.num_full_val_evals > 0
+
+    print("✅ RAG end-to-end optimization completed successfully!")
```

Comment:
Let's remove all prints from tests

Reply:
Removed all the print statements from the test.